aries_vcx/src/protocols/issuance_v2/formats/holder/hyperledger_indy.rs

```diff
@@ -0,0 +1,240 @@
+    ) -> VcxResult<HyperledgerIndyOfferDetails> {
+        let attachment = Self::extract_offer_attachment_content(offer_message)?;
+
+        Ok(serde_json::from_slice(&attachment)?)
+    }
+
+    async fn create_request_attachment_content(
+        offer_message: &OfferCredentialV2,
+        data: &Self::CreateRequestInput,
+    ) -> VcxResult<(Vec<u8>, HyperledgerIndyCreatedRequestMetadata)> {
+        let offer_bytes = Self::extract_offer_attachment_content(&offer_message)?;
+        let offer_payload = String::from_utf8(offer_bytes).map_err(|_| {
+            AriesVcxError::from_msg(
+                AriesVcxErrorKind::EncodeError,
+                "Expected payload to be a utf8 string",
+            )
+        })?;
+
+        let cred_def_id = parse_cred_def_id_from_cred_offer(&offer_payload)?;
+        let entropy = &data.my_pairwise_did;
+        let ledger = data.ledger;
+        let anoncreds = data.anoncreds;
+
+        let (credential_request, credential_request_metadata, _, credential_def_json) =
+            create_anoncreds_credential_request(
+                ledger,
+                anoncreds,
+                &cred_def_id,
+                &entropy,
+                &offer_payload,
+            )
+            .await?;
+
+        Ok((
+            credential_request.into(),
+            HyperledgerIndyCreatedRequestMetadata {
+                credential_request_metadata,
+                credential_def_json,
+            },
+        ))
+    }
+
+    async fn create_request_attachment_content_independent_of_offer(
+        _: &Self::CreateRequestInput,
+    ) -> VcxResult<(Vec<u8>, Self::CreatedRequestMetadata)> {
+        Err(AriesVcxError::from_msg(
+            AriesVcxErrorKind::ActionNotSupported,
+            "Anoncreds cannot create request payload independent of an offer",
+        ))
+    }
+
+    async fn process_and_store_credential(
+        issue_credential_message: &IssueCredentialV2,
+        user_input: &HyperledgerIndyStoreCredentialInput<R, A>,
+        request_metadata: &HyperledgerIndyCreatedRequestMetadata,
+    ) -> VcxResult<HyperledgerIndyStoredCredentialMetadata> {
+        let cred_bytes = Self::extract_credential_attachment_content(&issue_credential_message)?;
+        let credential_payload = String::from_utf8(cred_bytes).map_err(|_| {
+            AriesVcxError::from_msg(
+                AriesVcxErrorKind::EncodeError,
+                "Expected payload to be a utf8 string",
+            )
+        })?;
+
+        let ledger = user_input.ledger;
+        let anoncreds = user_input.anoncreds;
+
+        let rev_reg_id = _parse_rev_reg_id_from_credential(&credential_payload)?;
+        let rev_reg_def_json = if let Some(rev_reg_id) = rev_reg_id {
+            let json = ledger.get_rev_reg_def_json(&rev_reg_id).await?;
+            Some(json)
+        } else {
+            None
+        };
+
+        let cred_id = anoncreds
+            .prover_store_credential(
```

Comment:
Following comment is out of scope of this PR, non-actionable. The fact that we store credential here seems quite wrong.

1. Anoncreds API should not store anything, If anything, I would expect a `wallet` to be here.
2. Following up previous point, in fact, there should be no persistence here. This function should rather return an artifact, eg. some sort of credential data which shall be stored by consumer.

Our today's paradigm is that the protocol also end up representing "the artifact" is wrong and lead to application code, where credentials are represented by state machines in `Complete` states. 
The same problem we have with Connection protocol, (but afaik not as much with DidExchange), where the finalized state machine end up representing a didcomm channel (DID + Keys)

The problem with this coupling of 
- "output artifact" (be it "DID + Keys" or <credential in some format>) 
- with "state machine"

Will cause headaches whenever there is new version of protocol. In application code, there should be separate collection of artifacts (such as a didcomm channel, a credential of some type) and seaprate collection of state machines - really just helping out with tracking the ongoing conversations. The state machines should not become interface for working with underlying artifacts.

One more illustration of this problem is revocation. In V1 protocol, issuer state machine in states where a credential came to an existence had api `revoke_credential()`. Convenient but not part of aries protocol, just mean of controlling artifact created within an aries protocol.

Reply:
- Ultimately this is problem of applications built on top of `aries_vcx`. We should probably have a readme with some list of reccomendations, best practices for building on top of `aries_vcx`. 
- We could also deter implementations relying on protocols in final states by stripping off state machines of some APIs. For example, `Connection` in any state should not provide `send_message` functionality. `IssuerCredential` in final state should not provide `revoke` functionality.